### ISSUE TITLE
Unstable SAGE values

### ISSUE THREAD

**Issue by adityakumar04** [icc2115/sage#2]
Hi,

I ran SAGE on my test data which is 5k rows and 500 columns. The SAGE values that I get are very small (~0.003) and change direction and magnitude with subsequent runs. I also tried with both permutation sampling and iterated sampling however the difference is not very clear. Could you help me understand how to resolve this issue and how n_samples might affect my results.

**Comment by icc2115** [comment]
Hi Aditya,

Sorry for the confusion about the correct number of samples. The permutation sampling algorithm is the one described in the paper, and it tends to be a bit faster than the iterated sampling algorithm. I've just pushed a bunch of changes, and among other things it's now easy to get standard deviation error bars to accompany the SAGE value estimates. Check out the example notebooks ([credit.ipynb](https://github.com/icc2115/sage/blob/master/credit.ipynb) and [bike.ipynb](https://github.com/icc2115/sage/blob/master/bike.ipynb)) and let me know if they make sense.

The number of samples needs to be quite large in order to converge, although it's hard to say exactly how large for a given dataset. When working with a new dataset, two options for figuring out a sufficient number of samples are 1) keep increasing the number until the error bars seem small enough, 2) use the `detect_convergence` option to figure it out automatically. The downside to `detect_convergence` is that it might be conservative and calculate more samples than you really need.

Given the large number of columns in your dataset, a good starting point might be `num_permutations = 2 ** 16`, since that's about how many it took for our MNIST experiments to converge. Let me know how it goes, I'm happy to help troubleshoot more.

**Comment by adityakumar04** [comment]
Hi Ian, thanks a lot for your reply. Another concern I have is that when I pass classes as -1 and 1 then I get all SAGE values as negative however when I convert -1 class to 0 then I get positive SAGE values which actually make sense for my results. Random forest is agnostic to -1 and 1 as classes and gives same results. Can we include the same functionality in SAGE also.

**Comment by icc2115** [comment]
Hi Aditya, thanks for pointing this out. I had implemented the cross entropy loss function in a way that didn't accommodate (-1, 1) labeling, but I've just pushed some changes that fix it. Hopefully it works for you now!

**Comment by adityakumar04** [comment]
Thanks a lot Ian. This really helps a lot. I will continue using SAGE and see if I can get some more features added.
